### PR TITLE
balancer/randomsubsetting: Extend the unit tests in the randomsubsetting package

### PR DESCRIPTION
**Pull request by marek-szews**
Add more unit tests to increase test coverage of 'randomsubsetting' package.

RELEASE NOTES: none

### CODE REVIEW

**Comment by marek-szews** [comment]
This pull request is a continuation of the work started in [PR#8781](https://github.com/grpc/grpc-go/pull/8781). Due to a high volume of merge conflicts that made the previous PR impossible to merge cleanly, I have decided to close the old one and start fresh here with the latest changes.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 1]
```diff
@@ -259,3 +260,226 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+	} else {
+		reportBuilder.WriteString(fmt.Sprintf("Distribution is not uniform (χ²=%.2f > critical value=%.2f)", chi2, criticalValue))
+	}
+
+	return testPassed, reportBuilder
+}
+
+// ChiSquareCriticalValue calculates the critical value for alpha (e.g., 0.05)
+// and degrees of freedom (df).
+func chiSquareCriticalValue(alpha float64, df float64) float64 {
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go:465]: As mentioned [here](https://github.com/grpc/grpc-go/pull/8781#issuecomment-3962016617), we should try and see if this can reused here instead of reimplementing it.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 2]
```diff
@@ -259,3 +260,226 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+			want:       2,
+		},
+		{
+			endpoints:  makeEndpoints(150),
+			subsetSize: 50,
+			want:       50,
+		},
+	}
+
+	for _, tc := range testCases {
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go:287]: I don't see any point of adding a new test which will assert against the lenght of subset returned from `b.calculateSubset`.

Test `TestCalculateSubset_Simple` is sufficient, you just need to add a case when given endpoints are larger than the subset size.

Please remove `TestSubsettingEndpointsSimply` test.

**Review by Pranjali-2501 — COMMENTED** [review]
@marek-szews, PTAL at the comments.

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 2 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go:287]: Following your suggestions, I added a new case to the existing test in which the size of the subset is smaller than the number of Endpoints.

 FAIL: Test (0.06s)
    --- FAIL: Test/CalculateSubset_Simple (0.00s)
        --- FAIL: Test/CalculateSubset_Simple/SubsetSizeSmallerThanNumberOfEndpoints#01 (0.00s)
            /usr/local/google/home/szews/tmp/grpc-go/balancer/randomsubsetting/randomsubsetting_test.go:171: calculateSubset() returned diff (-want +got):
                  []resolver.Endpoint{
                + 	{Addresses: []resolver.Address{s`{Addr: "endpoint-13", ServerName: "", }`}},
                + 	{Addresses: []resolver.Address{s`{Addr: "endpoint-10", ServerName: "", }`}},
                + 	{Addresses: []resolver.Address{s`{Addr: "endpoint-12", ServerName: "", }`}},
                + 	{Addresses: []resolver.Address{s`{Addr: "endpoint-6", ServerName: "", }`}},
                  	{Addresses: {s`{Addr: "endpoint-0", ServerName: "", }`}},
                - 	{Addresses: []resolver.Address{s`{Addr: "endpoint-1", ServerName: "", }`}},
                - 	{Addresses: []resolver.Address{s`{Addr: "endpoint-2", ServerName: "", }`}},
                - 	{Addresses: []resolver.Address{s`{Addr: "endpoint-3", ServerName: "", }`}},
                - 	{Addresses: []resolver.Address{s`{Addr: "endpoint-4", ServerName: "", }`}},
                  }
FAIL
FAIL	google.golang.org/grpc/balancer/randomsubsetting	0.105s
FAIL
I have described this problem last time in details:
>>>>
Not exactly like that, function makeEndpoints(15) will produce the series of endpoints with a arithmetic order {endpoint_0, endpoint_1, ..., endpoint_14}. Same makeEndpoints(5) give us {endpoint_0, endpoint_1,.., endpoint_4}. But the result of LB.calculateSubset(eps) is unpredictable. One time we can received subset {endpoint_4, endpoint_5, endpoint_8, endpoint_11, endpoint_14} another time completely different. Validation of members of subset 99% will failed. I have left those two tests separated, because the first of them, check the content of set (boundary condition is meet so method always return the unchanged set - with a origin order) while the second test due to randomisation of content of subset, validate only the cardinality of the set (number of elements).
<<<<

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 1 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go:465]: If I have to reuse this code in randomsubsetting_test.go, then i need to do :

- Export it (rename it to PearsonsChiSquareTest).

- Move it to a more general test utility package if we want to follow strict package boundaries (since it is currently under a roundrobin directory).

Please let me know if the above proposal meets your expectations.

**Review by marek-szews — COMMENTED** [review]
Would you refer to my answers to the asked questions ?

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 3]
```diff
@@ -259,3 +284,123 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+
+// TestUniformDistributionOfEndpoints verifies that the random subsetting
+// policy achieves a uniform distribution across backends. From a set of N
+// numbers, it randomly selects K-times a subset of L numbers, where L < N.
+// Then it calculates how many times each number belonging to set N appears,
+// compute the variance and standard deviation, and use a Chi-Square test to
+// check whether the distribution is uniform.
+func (s) TestUniformDistributionOfEndpoints(t *testing.T) {
+	testCases := []struct {
+		eps        int
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go:300]: add a `name` field and add valuable names to all test cases.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 4]
```diff
@@ -259,3 +284,123 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+		subsetSize uint32
+		iteration  uint32
+		positive   bool
+	}{
+		{
+			eps:        16,
+			subsetSize: 4,
+			iteration:  10,
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go:309]: Why are we using such a small value for iterations here.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 5]
```diff
@@ -259,3 +284,123 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+			positive:   true,
+		},
+		{
+			eps:        8,
+			subsetSize: 4,
+			iteration:  3200,
+			positive:   true,
+		},
+		{
+			eps:        4,
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go:342]: When the subset size equals the total number of endpoints (eps == subsetSize), the algorithm returns all endpoints every time. 

We can remove this testcase.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 6]
```diff
@@ -259,3 +284,123 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+			subsetSize: 4,
+			iteration:  6400,
+			positive:   true,
+		},
+		{
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go:347]: I think we don't need the false assertion here. IMO, just having testcases that are expected to pass is sufficient.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 7]
```diff
@@ -259,3 +284,123 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+			subsetOfEndpoints := lb.calculateSubset(endpoints)
+
+			for _, ep := range subsetOfEndpoints {
+				EndpointCount[ep.Addresses[0].Addr]++
+			}
+		}
+
+		t.Logf("Test Case: Endpoints=%d, SubsetSize=%d, Iterations=%d", N, L, K)
+
+		observedCounts := make(map[string]float64)
```
Comment (Pranjali-2501) [inline balancer/randomsubsetting/randomsubsetting_test.go:393]: This can be done in a single iteration.
```
observedCounts := make(map[string]float64)
expectedCounts := make(map[string]float64)
for k, v := range EndpointCount {
	observedCounts[k] = float64(v)
	expectedCounts[k] = E
}
```

**Comment by Pranjali-2501** [comment]
@marek-szews , please fix [Testing / static checks (latest-1) (pull_request)](https://github.com/grpc/grpc-go/actions/runs/24768723582/job/72469315073?pr=9059).

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 6 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go:347]: The fact that your tests do not fail is actually the point—it confirms that the algorithm is operationally sound. However, checking for a uniform distribution serves a different purpose than simply checking for a "pass/fail" result.

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 5 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go:342]: I agree with your point regarding the first part of the statement; however, the successful outcome of the verification process is not entirely definitive. It primarily depends on the number of repetitions to ensure the sample is statistically representative.
The distribution will only begin to align once this specific threshold is exceeded. If we wish to examine the Load Balancer algorithm's actual output, a relevant test can be found in the TestCalculateSubset_Simple() function, specifically under the test name SubsetSizeEqualToNumberOfEndpoints.

It was my underlying objective to demonstrate the correlation between EPs, subsetSize, and the total iteration count. As subsetSize converges with EPs, the number of iterations required for successful verification increases significantly.

EndPoints              |16  |  8 | 4 |
subsetSize            |  4  |  4 | 4 |
repetition         1600 -> 3200 -> 6400

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 7 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go:393]: Done

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 3 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go:300]: Done

File: balancer/randomsubsetting/randomsubsetting_test.go (hunk 4 above)
Comment (marek-szews) [reply balancer/randomsubsetting/randomsubsetting_test.go:309]: That explains the situation perfectly! Using a hardcoded seed in your tests creates a deterministic environment where the algorithm's choices are fixed for every run.
While this is excellent for reproducibility, it can create a "false sense of randomness" if the specific seed happens to generate a sequence that fits your distribution criteria perfectly, even with low iterations.

**Review by marek-szews — COMMENTED** [review]
All of your comments have been reviewed and addressed. Please find my response below.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 8]
```diff
@@ -259,3 +284,128 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+
+		EndpointCount := make(map[string]int, N)
+		for _, ep := range endpoints {
+			EndpointCount[ep.Addresses[0].Addr] = 0
+		}
+
+		for i := 0; i < K; i++ {
+			lb := &subsettingBalancer{
+				cfg:        &lbConfig{SubsetSize: uint32(L)},
+				hashSeed:   uint64(i ^ 3 + K*i + L),
```
Comment (easwars) [inline balancer/randomsubsetting/randomsubsetting_test.go:381]: What is the significance of the way this seed value is computed? Why can't it just be `i`?

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 9]
```diff
@@ -259,3 +284,128 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+
+	for _, tc := range testCases {
+		endpoints := makeEndpoints(tc.eps)
+
+		// Theoretical Calculations
+		N := len(endpoints)
+		L := int(tc.subsetSize)
+		K := int(tc.iteration)
+		p := float64(L) / float64(N) // Probability of x ∈ N being drawn p(x) = L / N
+		E := float64(K) * p          // Expected Value (Mean) E(N) = K * p
```
Comment (easwars) [inline balancer/randomsubsetting/randomsubsetting_test.go:371]: Having single letter variable names that are spread through the function does not help with readability at all. Please get rid of these variables and instead use the RHS of these variables directly in the code.

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 10]
```diff
@@ -259,3 +284,128 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+
+		EndpointCount := make(map[string]int, N)
```
Comment (easwars) [inline balancer/randomsubsetting/randomsubsetting_test.go:373]: s/EndpointCount/endpointCount

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 11]
```diff
@@ -259,3 +284,128 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
...
+		t.Logf("Test Case: Endpoints=%d, SubsetSize=%d, Iterations=%d", N, L, K)
+
+		observedCounts := make(map[string]float64)
+		expectedCounts := make(map[string]float64)
+		for k, v := range EndpointCount {
+			observedCounts[k] = float64(v)
+			expectedCounts[k] = E
+		}
+
+		err := roundrobin.PearsonsChiSquareTest(t, observedCounts, expectedCounts, 0.05)
```
Comment (easwars) [inline balancer/randomsubsetting/randomsubsetting_test.go:400]: How was the value of `0.05` arrived at?

File: balancer/randomsubsetting/randomsubsetting_test.go [hunk 12]
```diff
@@ -259,3 +284,128 @@ func (s) TestSubsettingBalancer_DeterministicSubset(t *testing.T) {
 		t.Fatal("Timed out waiting for second child policy update")
 	}
 }
+
+// TestUniformDistributionOfEndpoints verifies that the random subsetting
+// policy achieves a uniform distribution across backends. From a set of N
+// numbers, it randomly selects K-times a subset of L numbers, where L < N.
+// Then it calculates how many times each number belonging to set N appears,
+// compute the variance and standard deviation, and use a Chi-Square test to
+// check whether the distribution is uniform.
```
Comment (easwars) [inline balancer/randomsubsetting/randomsubsetting_test.go:293]: From what I can see, this test is actually testing the hashing function. If we really want to test the functionality in the policy, we should be overriding the hashing function in some deterministic fashion, so that we can verify the algorithm in the policy.

I'm concerned about a bunch of things in this test, but most importantly, if this test fails, how would one go about debugging? What do the scenarios in the different table entries correspond to, how were they arrived at?
